propagation/w3c/src/main/java/brave/propagation/w3c/TraceparentFormat.java

```diff
@@ -0,0 +1,297 @@
+          default:
+            throw new AssertionError();
+        }
+
+        buffer = 0L;
+        currentFieldLength = 0;
+        continue;
+      }
+
+      // At this point, 'c' is not a hyphen
+
+      // When we get to a non-hyphen at position 16, we have a 128-bit trace ID.
+      if (currentField == FIELD_TRACE_ID && currentFieldLength == 16) {
+        // traceIdHigh can be zeros when a 64-bit trace ID is encoded in 128-bits.
+        traceIdHighZero = buffer == 0L;
+        builder.traceIdHigh(buffer);
+
+        // This character is the next hex. If it isn't, the next iteration will throw. Either way,
+        // reset so that we can capture the next 16 characters of the trace ID.
+        buffer = 0L;
+      }
+
+      currentFieldLength++;
+
+      // The rest of this is normal lower-hex decoding
+      buffer <<= 4;
+      if (c >= '0' && c <= '9') {
+        buffer |= c - '0';
+      } else if (c >= 'a' && c <= 'f') {
+        buffer |= c - 'a' + 10;
+      } else {
+        log(currentField, "Invalid input: only valid characters are lower-hex for {0}");
+        return null;
+      }
+    }
+
+    try {
+      return builder.build();
+    } catch (IllegalArgumentException e) { // trace ID or span ID were all zeros
+      Platform.get().log(e.getMessage(), null);
+      return null;
+    }
+  }
+
+  static boolean validateFieldLength(int field, int length) {
+    int expectedLength =
+      (field == FIELD_TRACE_FLAGS || field == FIELD_VERSION) ? 2
+        : field == FIELD_TRACE_ID ? 32 : 16;
+    if (length == 0) {
+      log(field, "Invalid input: empty {0}");
+      return false;
+    } else if (length < expectedLength) {
+      log(field, "Invalid input: {0} is too short");
+      return false;
+    } else if (length > expectedLength) {
+      log(field, "Invalid input: {0} is too long");
+      return false;
+    }
+    return true;
+  }
+
+  static void logReadAllZeros(int currentField) {
+    log(currentField, "Invalid input: read all zeros {0}");
+  }
+
+  static void log(int fieldCode, String s) {
+    String field;
+    switch (fieldCode) {
+      case FIELD_VERSION:
+        field = "version";
+        break;
+      case FIELD_TRACE_ID:
+        field = "trace ID";
+        break;
+      // Confusingly, the spec calls the span ID field parentId
+      // https://w3c.github.io/trace-context/#parent-id
+      case FIELD_PARENT_ID:
+        field = "parent ID";
+        break;
+      case FIELD_TRACE_FLAGS:
+        field = "trace flags";
+        break;
+      default:
+        throw new AssertionError("field code unmatched: " + fieldCode);
+    }
+    Platform.get().log(s, field, null);
+  }
+
+  static byte[] asciiToNewByteArray(char[] buffer, int length) {
+    byte[] result = new byte[length];
+    for (int i = 0; i < length; i++) {
+      result[i] = (byte) buffer[i];
+    }
+    return result;
+  }
+
+  static final ThreadLocal<WeakReference<TraceContext.Builder>> BUILDER_REF = new ThreadLocal<>();
```

Comment:
TODO: I've not analyzed benchmarks with vs without, yet, but it should help reduce allocations without resorting to internal code. Also, we need to add the classloader test which would break if you switch to non weak ref.

Reply:
it sincerely didn't help